Mathlib/MeasureTheory/Measure/Haar/Unique.lean

```diff
@@ -282,56 +302,466 @@ lemma integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport
-    · exact IsOpen.measure_pos μ' g_cont.isOpen_support ⟨1, g_one⟩
+    · exact IsOpen.measure_pos ν g_cont.isOpen_support ⟨1, g_one⟩
     · exact g_cont.integrable_of_hasCompactSupport g_comp
-  have := integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ' μ g_cont g_comp
-  simp only [H, zero_smul, integral_zero_measure] at this
-  linarith
+  change (haarScalarFactor μ' ν : ℝ) * ∫ (x : G), g x ∂ν =
+    (haarScalarFactor μ' μ * haarScalarFactor μ ν : ℝ≥0) * ∫ (x : G), g x ∂ν at Z
+  simpa only [mul_eq_mul_right_iff (M₀ := ℝ), int_g_pos.ne', or_false, NNReal.eq_iff] using Z
+
+@[to_additive (attr := simp)]
+lemma haarScalarFactor_self (μ : Measure G) [IsHaarMeasure μ] :
+    haarScalarFactor μ μ = 1 := by
+  -- The group has to be locally compact, otherwise the scalar factor is 1 by definition.
+  by_cases hG : LocallyCompactSpace G; swap
+  · simp [haarScalarFactor, hG]
+  -- Fix some nonzero continuous function with compact support `g`.
+  obtain ⟨⟨g, g_cont⟩, g_comp, g_nonneg, g_one⟩ :
+    ∃ (g : C(G, ℝ)), HasCompactSupport g ∧ 0 ≤ g ∧ g 1 ≠ 0 := exists_continuous_nonneg_pos 1
+  have Z := integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ μ g_cont g_comp
+  have int_g_pos : 0 < ∫ x, g x ∂μ := by
+    apply (integral_pos_iff_support_of_nonneg g_nonneg _).2
+    · exact IsOpen.measure_pos μ g_cont.isOpen_support ⟨1, g_one⟩
+    · exact g_cont.integrable_of_hasCompactSupport g_comp
+  rw [integral_smul_nnreal_measure, eq_comm] at Z
+  change (haarScalarFactor μ μ : ℝ) * ∫ (x : G), g x ∂μ = ∫ (x : G), g x ∂μ at Z
+  simpa [mul_eq_right₀ (M₀ := ℝ), int_g_pos.ne'] using Z
+
+  /-- The scalar factor between two left-invariant measures is non-zero when both measures are
+positive on open sets. -/
+@[to_additive]
+lemma haarScalarFactor_pos_of_isHaarMeasure (μ' μ : Measure G) [IsHaarMeasure μ]
+    [IsHaarMeasure μ'] : 0 < haarScalarFactor μ' μ :=
+  pos_iff_ne_zero.2 (fun H ↦ by simpa [H] using haarScalarFactor_eq_mul μ' μ μ')
+
+@[deprecated] alias haarScalarFactor_pos_of_isOpenPosMeasure :=
+  haarScalarFactor_pos_of_isHaarMeasure  -- deprecated on 2024-02-12
+
+@[deprecated] alias addHaarScalarFactor_pos_of_isOpenPosMeasure :=
+  addHaarScalarFactor_pos_of_isAddHaarMeasure  -- deprecated on 2024-02-12
+
+/-!
+### Uniqueness of measure of sets with compact closure
+
+Two left invariant measures give the same measure to sets with compact closure, up to the
+scalar `haarScalarFactor μ' μ`.
+
+This is a tricky argument, typically not done in textbooks (the textbooks version all require one
+version of regularity or another). Here is a sketch, based on
+McQuillan's answer at https://mathoverflow.net/questions/456670/.
+
+Assume for simplicity that all measures are normalized, so that the scalar factors are all `1`.
+First, from the fact that `μ` and `μ'` integrate in the same way compactly supported functions,
+they give the same measure to compact "zero sets", i.e., sets of the form `f⁻¹ {1}`
```

Comment:
maybe change "zero sets" to "level sets" (and 'f-1{1}' to 'f-1{a} for a neq 0').

Reply:
Zero set is unfortunately a standard terminology in this corner of the litterature. That's why I am putting it in the docstring, to make sure that people looking for this can locate the relevant lemmas.